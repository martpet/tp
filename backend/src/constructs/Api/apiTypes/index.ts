export * from './EndpointsCookies';
export * from './EndpointsEnvironments';
export * from './EndpointsHeaders';
export * from './EndpointsOptions';
export * from './EndpointsQueryStrings';
export * from './OauthCookieProps';
export * from './OauthTokens';
