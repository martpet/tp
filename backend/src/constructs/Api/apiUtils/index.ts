export * from './cookieName';
export * from './errorResponse';
export * from './handlerWrapper';
export * from './itResolvesWithApiError';
