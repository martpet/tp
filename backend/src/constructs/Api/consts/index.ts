export * from './authorizationHeader';
export * from './defaultCachePolicyProps';
export * from './defaultSecurityHeadersBehavior';
export * from './lambdaEdgeViewerEvent';
