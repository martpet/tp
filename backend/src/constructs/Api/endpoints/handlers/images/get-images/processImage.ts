import { SdkStreamMixin } from '@aws-sdk/types';
import { DetectModerationLabelsCommand, RekognitionClient } from 'lambda-layer';
import sharp from 'sharp';

const rekognition = new RekognitionClient({});

type Props = {
  stream: SdkStreamMixin;
  photoBucket: string;
  s3ObjectName: string;
  quality?: number;
  size?: number;
};

export async function processImage({
  stream,
  photoBucket,
  s3ObjectName,
  size = 1500,
  quality = 100,
}: Props) {
  const moderationCommand = new DetectModerationLabelsCommand({
    Image: { S3Object: { Bucket: photoBucket, Name: s3ObjectName } },
    MinConfidence: 90,
  });
  const { ModerationLabels } = await rekognition.send(moderationCommand);
  const uint8 = await stream.transformToByteArray();
  const image = sharp(uint8).jpeg({ quality }).resize(size);
  const isModerated = ModerationLabels?.some(({ Name }) => Name === 'Explicit Nudity');

  if (isModerated) {
    image.blur(75);
  }

  const buffer = await image.toBuffer();
  const base64 = buffer.toString('base64');
  return base64;
}
