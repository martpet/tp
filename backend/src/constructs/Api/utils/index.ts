export * from './cookieName';
export * from './errorResponse';
export * from './getIdTokenPayload';
export * from './lambdaEdgeErrorResponse';
export * from './parseEventCookies';
export * from './parseLambdaEdgeEventCookies';
export * from './testHelpers';
