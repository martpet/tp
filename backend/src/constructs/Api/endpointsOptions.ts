import { Auth, Tables } from '~/constructs';
import { EnvName } from '~/types';

import { EndpointsOptions } from './apiTypes/EndpointsOptions';

type EndpointsOptionsProps = {
  auth: Auth;
  tables: Tables;
  envName: EnvName;
};

export const endpointsOptions = ({
  tables,
  auth,
  envName,
}: EndpointsOptionsProps): EndpointsOptions => {
  const { authDomain, loginCallbackUrl, logoutCallbackUrl, logoutCallbackLocalhostUrl } =
    auth;
  const clientId = auth.userPoolClient.userPoolClientId;
  const { sessionsTable } = tables;

  return {
    '/login': {
      cookies: ['oauth'],
      queryStrings: ['provider'],
      methods: {
        GET: {
          environment: { authDomain, clientId, loginCallbackUrl },
        },
      },
    },

    '/loginCallback': {
      cookies: ['oauth'],
      queryStrings: ['code', 'state'],
      methods: {
        GET: {
          environment: { authDomain, clientId, loginCallbackUrl, envName },
          cb: (handler) => sessionsTable.grantWriteData(handler),
        },
      },
    },

    '/logout': {
      cookies: ['session'],
      headers: ['referer'],
      methods: {
        GET: {
          environment: {
            authDomain,
            clientId,
            logoutCallbackUrl,
            logoutCallbackLocalhostUrl,
            envName,
          },
          cb: (handler) => sessionsTable.grantWriteData(handler),
        },
      },
    },
  };
};
