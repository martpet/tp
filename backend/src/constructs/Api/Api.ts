import { HttpApi } from '@aws-cdk/aws-apigatewayv2-alpha';
import { Fn, NestedStack, RemovalPolicy } from 'aws-cdk-lib';
import { Distribution } from 'aws-cdk-lib/aws-cloudfront';
import { HttpOrigin } from 'aws-cdk-lib/aws-cloudfront-origins';
import { ARecord, RecordTarget } from 'aws-cdk-lib/aws-route53';
import { CloudFrontTarget } from 'aws-cdk-lib/aws-route53-targets';
import { Bucket } from 'aws-cdk-lib/aws-s3';
import { Construct } from 'constructs';

import { Auth, Photos, Tables, Zone } from '~/constructs';
import { apiSubdomain, appEnvs } from '~/consts';
import { getEnvName } from '~/utils';

import { createDistroBehaviors } from './endpoints/createDistroBehaviors';
import { createRoutes } from './endpoints/createRoutes';

type Props = {
  zone: Zone;
  auth: Auth;
  tables: Tables;
  photos: Photos;
};

export class Api extends NestedStack {
  constructor(scope: Construct, { zone, tables, auth, photos }: Props) {
    super(scope, 'Api');

    const envName = getEnvName(this);
    const { appDomain } = appEnvs[envName];
    const apiDomain = `${apiSubdomain}.${appDomain}`;
    const api = new HttpApi(this, 'http-api');
    const origin = new HttpOrigin(Fn.select(1, Fn.split('://', api.apiEndpoint)));

    const logBucket = new Bucket(this, 'api-distro-log-bucket', {
      removalPolicy: RemovalPolicy.DESTROY,
      autoDeleteObjects: true,
    });

    const distribution = new Distribution(this, 'api-distribution', {
      defaultBehavior: { origin },
      domainNames: [apiDomain],
      certificate: zone.certificate,
      enableLogging: true,
      logIncludesCookies: true,
      logBucket,
    });

    createRoutes({ scope, api, auth, tables, photos });

    createDistroBehaviors({
      scope,
      api,
      auth,
      tables,
      distribution,
      origin,
    });

    new ARecord(this, 'api-a-record', {
      zone: zone.hostedZone,
      recordName: apiSubdomain,
      target: RecordTarget.fromAlias(new CloudFrontTarget(distribution)),
    });
  }
}
