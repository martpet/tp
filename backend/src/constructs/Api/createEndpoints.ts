import { HttpApi, HttpMethod } from '@aws-cdk/aws-apigatewayv2-alpha';
import { HttpLambdaIntegration } from '@aws-cdk/aws-apigatewayv2-integrations-alpha';
import {
  CachePolicy,
  Distribution,
  OriginRequestCookieBehavior,
  OriginRequestHeaderBehavior,
  OriginRequestPolicy,
  OriginRequestPolicyProps,
  OriginRequestQueryStringBehavior,
} from 'aws-cdk-lib/aws-cloudfront';
import { HttpOrigin } from 'aws-cdk-lib/aws-cloudfront-origins';
import { NodejsFunctionProps } from 'aws-cdk-lib/aws-lambda-nodejs';
import { Construct } from 'constructs';
import { Writable } from 'type-fest';

import { capitalize, createNodejsFunction } from '~/utils';

import { EndpointsOptions, MethodOptions } from './apiTypes';

type CreateEndpointProps = {
  scope: Construct;
  distribution: Distribution;
  api: HttpApi;
  origin: HttpOrigin;
  options: EndpointsOptions;
};

export const createEndpoints = ({
  scope,
  distribution,
  api,
  origin,
  options,
}: CreateEndpointProps) => {
  Object.entries(options).forEach(
    ([path, { methods, cookies, headers, queryStrings }]) => {
      addDistroBehavior({
        scope,
        distribution,
        origin,
        path,
        headers,
        cookies,
        queryStrings,
      });

      Object.entries(methods).forEach(([method, methodOptions]) => {
        addApiRoute({ scope, api, path, method, methodOptions });
      });
    }
  );
};

type AddDistroBehaviorProps = {
  scope: Construct;
  distribution: Distribution;
  origin: HttpOrigin;
  path: string;
  headers?: string[];
  cookies?: string[];
  queryStrings?: Readonly<string[]>;
};

function addDistroBehavior({
  scope,
  distribution,
  origin,
  path,
  cookies,
  headers,
  queryStrings,
}: AddDistroBehaviorProps) {
  const pathName = path.replace('/', '');
  const policyProps: Writable<OriginRequestPolicyProps> = {};

  if (headers) {
    policyProps.headerBehavior = OriginRequestHeaderBehavior.allowList(...headers);
  }

  if (cookies) {
    policyProps.cookieBehavior = OriginRequestCookieBehavior.allowList(...cookies);
  }

  if (queryStrings) {
    policyProps.queryStringBehavior = OriginRequestQueryStringBehavior.allowList(
      ...queryStrings
    );
  }

  distribution.addBehavior(path, origin, {
    cachePolicy: CachePolicy.CACHING_DISABLED,
    originRequestPolicy: new OriginRequestPolicy(scope, pathName, policyProps),
  });
}

type AddApiRouteProps = {
  scope: Construct;
  api: HttpApi;
  path: string;
  method: string;
  methodOptions: MethodOptions<unknown>;
};

function addApiRoute({ scope, api, path, method, methodOptions }: AddApiRouteProps) {
  const pathName = path.replace('/', '');
  const methodName = method.toLowerCase();
  const fileName = `${methodName}-${pathName}`;
  const id = `${capitalize(methodName)}${capitalize(fileName)}`;
  const { environment, cb } = methodOptions;
  const handlerProps: Writable<NodejsFunctionProps> = {
    entry: `${__dirname}/handlers/${pathName}/${fileName}/${fileName}.ts`,
    environment,
  };

  const handler = createNodejsFunction(scope, `${id}Handler`, handlerProps);

  if (cb) {
    cb(handler);
  }

  api.addRoutes({
    path,
    methods: [method as HttpMethod],
    integration: new HttpLambdaIntegration(`${id}Integration`, handler),
  });
}
