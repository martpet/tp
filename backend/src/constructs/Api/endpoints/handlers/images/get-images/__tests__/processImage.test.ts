import {
  DetectModerationLabelsCommand,
  RekognitionClient,
} from '@aws-sdk/client-rekognition';
import { mockClient } from 'aws-sdk-client-mock';
import sharp from 'sharp';

import { itCalls, itResolves, itSendsAwsCommand } from '~/constructs/Api/utils';

import { processImage } from '../processImage';

vi.mock('sharp');

const rekognitionMock = mockClient(RekognitionClient);

beforeEach(() => {
  rekognitionMock.reset();
  rekognitionMock.on(DetectModerationLabelsCommand).resolves({});
});

const args = [
  {
    stream: {
      transformToByteArray: () => Promise.resolve('dummyByteArray'),
    },
    photoBucket: 'dummyPhotoBucketName',
    s3ObjectName: 'dummyS3ObjectName',
    size: 500,
    quality: 80,
  },
] as unknown as Parameters<typeof processImage>;

describe('processImage', async () => {
  itSendsAwsCommand(DetectModerationLabelsCommand, rekognitionMock, processImage, args);
  itCalls(sharp, processImage, args);
  itCalls(sharp().jpeg, processImage, args);
  itCalls(sharp().resize, processImage, args);
  itCalls(sharp().toBuffer, processImage, args);
  itCalls((await sharp().toBuffer()).toString, processImage, args);
  itResolves(processImage, args);

  describe('when optional props are missing', () => {
    const propsClone = { ...args[0] };
    const argsClone = [propsClone];
    propsClone.size = undefined;
    propsClone.quality = undefined;
    itCalls(sharp().jpeg, processImage, argsClone);
    itCalls(sharp().resize, processImage, argsClone);
  });

  describe('when "DetectModerationLabelsCommand" returns "Explicit Nudity"', () => {
    beforeEach(() => {
      rekognitionMock.on(DetectModerationLabelsCommand).resolves({
        ModerationLabels: [{ Name: 'Explicit Nudity' }],
      });
    });

    itCalls(sharp().blur, processImage, args);
  });
});
