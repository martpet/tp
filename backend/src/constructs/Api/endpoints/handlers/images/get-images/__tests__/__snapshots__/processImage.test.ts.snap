// Vitest Snapshot v1

exports[`processImage > calls "jpeg" with correct args 1`] = `
[
  [
    {
      "quality": 80,
    },
  ],
]
`;

exports[`processImage > calls "resize" with correct args 1`] = `
[
  [
    500,
  ],
]
`;

exports[`processImage > calls "sharp" with correct args 1`] = `
[
  [
    "dummyByteArray",
  ],
]
`;

exports[`processImage > calls "toBuffer" with correct args 1`] = `
[
  [],
]
`;

exports[`processImage > calls "toString" with correct args 1`] = `
[
  [
    "base64",
  ],
]
`;

exports[`processImage > resolves with a correct value 1`] = `"dummyBufferToString"`;

exports[`processImage > sends "DetectModerationLabelsCommand" from "RekognitionClient" with correct args 1`] = `
{
  "Image": {
    "S3Object": {
      "Bucket": "dummyPhotoBucketName",
      "Name": "dummyS3ObjectName",
    },
  },
  "MinConfidence": 90,
}
`;

exports[`processImage > when "DetectModerationLabelsCommand" returns "Explicit Nudity" > calls "blur" with correct args 1`] = `
[
  [
    75,
  ],
]
`;

exports[`processImage > when optional props are missing > calls "jpeg" with correct args 1`] = `
[
  [
    {
      "quality": 100,
    },
  ],
]
`;

exports[`processImage > when optional props are missing > calls "resize" with correct args 1`] = `
[
  [
    1500,
  ],
]
`;
